src/main/java/org/opensearch/neuralsearch/plugin/query/NeuralQueryBuilder.java

```diff
@@ -207,12 +212,10 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
                     actionListener.onResponse(null);
                 }, actionListener::onFailure)))
             );
-            vectorSupplier(vectorSetOnce::get);
+            return new NeuralQueryBuilder(fieldName(), queryText(), modelId(), k(), vectorSetOnce::get);
         }
 
-        // Rewrites will continuously happen until the supplier is set and the vector is generated. Rewrites will stop
-        // once this object is returned. Hence, if we get here, we need to return a new object
-        return new NeuralQueryBuilder(fieldName(), queryText(), modelId(), k(), vectorSupplier());
+        return this;
```

Comment:
I think the third commit fixed the rescore exception of too many rewrite rounds. The key part is no need to return a new NeuralQueryBuilder. That's why the [PoC code](https://github.com/ylwu-amzn/k-NN/commit/d1164974695fe85186d4f0b5f40c5eebfa16e2a7#diff-19fa1b5328b894071230d33a4d215f96ec7cefd853500baf7b770b0fddf53ca6R172) worked for rescore before

Reply:
Thanks for fixing the rescore issue. The code LGTM. 
Please also test other possible use cases.